roborl/ddpg/noise.py

```diff
@@ -0,0 +1,44 @@
+import numpy as np
+import torch
+
+from torch.autograd import Variable
+
+class ParamNoise:
+
+    def __init__(self, batch_size, memory, sigma=0.1, delta=0.1, alpha=1.01):
+        self.alpha = alpha
+        self.batch_size = batch_size
+        self.memory = memory
+        self.sigma = sigma
+        self.delta = delta
+        self.orig_sigma = sigma
+
+    def distance(self, model, perturbed_model):
+        ''' DDPG Distance '''
+        batch = self.memory.sample_batch(self.batch_size)
+        return torch.sqrt(torch.mean((model(batch.states) - perturbed_model(batch.states))**2))
+
+    def update_sigma(self, model, perturbed_model):
+        '''Updates sigma at each timestep'''
+        if len(self.memory) < self.batch_size:
+            return
+        dist = self.distance(model, perturbed_model)
+        if dist.data.numpy() <= self.delta:
+            self.sigma *= self.alpha
+        else:
+            self.sigma *= 1 / self.alpha
+    
+    def perturb_model(self, model):
+        perturbed_model = model.clone()
+        for param in perturbed_model.parameters():
+            noise = torch.normal(means=torch.zeros_like(param), std=self.sigma)
+            param.add(noise)
```

Comment:
I think this should be `param.add_(noise)`, otherwise it is not in-place

Reply:
I get a runtime error with the in-place addition:

`< RuntimeError: a leaf Variable that requires grad has been used in an in-place operation. >`

Reply:
oh yeah I read about that. Then you probably need to get the `state_dict` add the noise and load it back into the model, because in this way you don't change the model parameters. `param.add(noise)` returns a new variable which is ignored in this case.

Reply:
or maybe `detach()` them, because we never need the gradients wrt. to these params.

http://pytorch.org/docs/master/autograd.html#in-place-operations-on-variables
http://pytorch.org/docs/master/autograd.html#torch.autograd.Variable.detach